### PR TITLE
Use _mm256_rsqrt_ps for rsqrt on CPU for float instead of using division and _mm256_sqrt_ps.

### PR DESCRIPTION
**Pull request by xuhdev**
Currently for single-precision floating-point number, division and
_mm256_sqrt_ps are used. While the current method is more accurate,
using _mm256_rsqrt_ps is faster. See
https://github.com/pytorch/pytorch/pull/8488#discussion_r195911562 .

This commit switches to _mm256_rsqrt_ps, because:

- This is more consistent with the CUDA implementation, which uses
  rsqrt, which is faster but less accurate:
  https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#standard-functions

- Users who call torch.rsqrt() are more likely to seek for speed than
  accuracy. Otherwise, they can use 1/torch.sqrt().

Another possibility would be to add an option to decide which version to
use, as rsqrt has a faster derivative.

Test on small3d is also removed, because test_rsqrt is sufficiently comprehensive, and small3d does not allow specifying max relative error.

Benchmark (RHEL 7.3, Release, g++ 8.3, turbo off):

```python
import timeit

for n, t in [(10_000, 800000),
             (100_000, 80000)]:
    print(f'a.rsqrt() (a.numel() == {n}) for {t} times')
    print(timeit.timeit(f'a.rsqrt()',
                        setup=f'import torch; a = torch.arange({n}, device="cpu", dtype=torch.float)',
                        number=t))
```

Before:

```
a.rsqrt() (a.numel() == 10000) for 800000 times
6.847818954000104
a.rsqrt() (a.numel() == 100000) for 80000 times
4.606836445998852
```

After:

```
a.rsqrt() (a.numel() == 10000) for 800000 times
3.61814929799948
a.rsqrt() (a.numel() == 100000) for 80000 times
2.1390665140006604
```

### CODE REVIEW

**Comment by xuhdev** [comment]
@vishwakftw @SsnL I added benchmarking results in the original comment: rsqrt is way faster than 1/sqrt, but it guarantees less accuracy.

**Comment by cpuhrsch** [comment]
Please update the documentation for rsqrt to document this very important fact including the work-around you wrote.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Documentation

**Comment by xuhdev** [comment]
@cpuhrsch Updated. Note that I took out `reciprocal()` in the workaround, because it seems like reciprocal itself suffers from an issue similar to this one...

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -4624,10 +4624,15 @@ def merge_dicts(*dicts):
...
 
 Returns a new tensor with the reciprocal of the square-root of each of
-the elements of :attr:`input`.
+the elements of ``input``.
 
 .. math::
     \text{out}_{i} = \frac{1}{\sqrt{\text{input}_{i}}}
+
+Note that this function, whenever possible, attempts to use an algorithm that
+is faster but guarantees less accuracy than that of ``1/torch.sqrt(input)``. If
```
Comment (gchanan) [inline torch/_torch_docs.py]: nit: I think it's clearer if you say "it does not guarantee as much accuracy" -- there are almost certainly cases where the the accuracy is the same.

**Comment by albanD** [comment]
Quick test using master before this PR:

```python
>>> a = torch.rand(1000).mul(0.001) # Get a bunch of small nunbers (error is smaller with bigger numbers)
>>> a.rsqrt().sub(a.cuda().rsqrt().cpu()).abs().max() # Max difference between the cpu and cuda version
tensor(0.0005)
>>> (1/a.sqrt()).sub((1/a.cuda().sqrt()).cpu()).abs().max() # Max difference with 1/a.sqrt()
tensor(0.)
```

@xuhdev can you confirm that with this diff, the cpu/cuda behavior with `rsqrt()` becomes the same?

**Comment by xuhdev** [comment]
No it's not the same (I changed your test to relative error because that's what really captures the accuracy in floating point computation):

```python
>>> (a.rsqrt().sub(a.cuda().rsqrt().cpu())/a.rsqrt()).abs().max()
tensor(0.0003)
>>> ((1/a.sqrt()).sub((1/a.cuda().sqrt()).cpu())/(1/a.sqrt())).abs().max()
tensor(0.)
```

I figured that [_mm256_rsqrt_ps](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_rsqrt_ps&expand=4804) produces a maximum relative error of `1.5*2^-12` (0.000366), and the [CUDA rsqrtf produces a maximum relative error](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html) of 2ulp (2.38e-7). Seems like both employs approximation but differ in the extent of approximation.

**Review by albanD — APPROVED** [review]

**Review by albanD — CHANGES_REQUESTED** [review]
The test failures are valid. Could you fix them please?

**Comment by xuhdev** [comment]
I've pushed a fix. Admittedly this is really much faster, but may also bring compatibility issues with past code. Do we want a `fast` option that defaults to `False` that corresponds to old CPU behavior and a more accurate CUDA behavior? My major concern is that using `rsqrt` can bring the benefit of a more optimized derivative than `1/sqrt`.

**Comment by albanD** [comment]
Well the gradient computation is defined here independently of the implementation (even though it depends on the output value): https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L732-L733

In this case, I think it would be safer to have the precise version by default (given that the fast cpu version trips our current tests) and add a `fast` argument that enables the fast version.
This will slow down current cuda implementations, @ngimel  do you think it is worth doing?

**Comment by ngimel** [comment]
I'm sorry, what is the exact question here? I agree with the points made above that if someone is using `rsqrt` they are probably ok with lower accuracy. Gradient formula is not really an issue I think - if result was imprecise, gradient is based off the imprecise result and is also imprecise, and in any case getting better accuracy for gradient would require recomputing the forward result in the first place, which is never what we want.

**Review by albanD — CHANGES_REQUESTED** [review]
Ok, after looking into this more, tf use unconditionally the fast, imprecise version. So I think it is better for us to do this change to use `rsqrt`.
I think you missed other places in the code base that still use the old implementation [here](https://github.com/pytorch/pytorch/blob/587996ef0479ec5ab2fcb26e78eacb4c56ad171a/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L319) and [here](https://github.com/pytorch/pytorch/blob/fdcb203e8ef18011512a55974918ee8c62637f5b/aten/src/TH/THMath.h#L17) and [here](https://github.com/pytorch/pytorch/blob/fdcb203e8ef18011512a55974918ee8c62637f5b/aten/src/TH/THMath.h#L25) and [here](https://github.com/pytorch/pytorch/blob/fdcb203e8ef18011512a55974918ee8c62637f5b/aten/src/ATen/cpu/vec256/vec256_double.h#L189) and [here](https://github.com/pytorch/pytorch/blob/fdcb203e8ef18011512a55974918ee8c62637f5b/aten/src/ATen/cpu/vec256/vec256_base.h#L323). These should be changed to have consistent results for different types and cpu capabilities.
Also for the jit team do we want to change the version in jit templates [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/fuser/cpu/resource_strings.h#L38-L44) (cc @suo ).

**Comment by albanD** [comment]
@ngimel  the exact question was: "do we want to change the behavior of the `rsqrt()` function?" Currently the cpu version does `1/sqrt(x)` while the gpu version does `rsqrt(x)`.

**Comment by ngimel** [comment]
Yes, we should, so that both cpu and gpu use fast approximation, which amounts to changing behavior on cpu as this PR suggests, and leaving gpu behavior as is.

**Comment by xuhdev** [comment]
@albanD I deleted two of your pointed rsqrts in #29688 .

The fast version of `rsqrt` does not exist in all cases. For example, on CUDA, `rsqrt` for double is the same as `1/sqrt`. There is also no corresponding AVX* instruction for double. What rsqrt does is to be fast if possible, but guarantees less accuracy.

**Review by albanD — APPROVED** [review]
@xuhdev sounds good. Thanks for the update. Let's wait for the CI to be back and then we can merge this.

**Review by gchanan — COMMENTED** [review]
so we are doing different precision calculations based on if we select a vectorized dispatch or not?  That seems pretty surprising.  Are there other cases where we do that?

**Review by gchanan — CHANGES_REQUESTED** [review]
if my interpretation is correct that we do different calculations on contiguous and not, I don't think we should commit this.  This can end up in surprising results that are difficult to track down.

**Comment by xuhdev** [comment]
@gchanan
> so we are doing different precision calculations based on if we select a vectorized dispatch or not?

Yes, precision would be different.

> That seems pretty surprising. Are there other cases where we do that?

No not what I know.

What is your suggestion then? I understand your concern, but it looks like a pretty difficult question though, as long as we want the fast rsqrt to be available; I see eigen uses the [more accurate implementation for scalar rsqrt](https://github.com/eigenteam/eigen-git-mirror/blob/633da008a45378f161a4e5dcb4442faf8b9b5e84/Eigen/src/Core/functors/UnaryFunctors.h#L393), but for packed they use [the less accurate one when available](https://github.com/eigenteam/eigen-git-mirror/blob/633da008a45378f161a4e5dcb4442faf8b9b5e84/Eigen/src/Core/arch/AVX/MathFunctions.h#L91).

Maybe a `fast` option, as I mentioned earlier, is a good compromise we can do?
